[Test Coverage] [SVG2] Allow leading and trailing whitespace in svg attributes using <angle>

Ahmad-S792 opened the pull request:
#### 2fa7080d7d85fe35d18be308ec6e2b794ed5afac
<pre>
[Test Coverage] [SVG2] Allow leading and trailing whitespace in svg attributes using &lt;angle&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=261656">https://bugs.webkit.org/show_bug.cgi?id=261656</a>
<a href="https://rdar.apple.com/115963075">rdar://115963075</a>

Reviewed by Taher Ali.

This patch adds test coverage for whitespace handling in SVG attributes
using &lt;angle&gt;, matching Gecko / Firefox and Blink / Chromium behavior.

Merge: <a href="https://github.com/chromium/chromium/commit/77c8625da85bd956992b81233d451b87c39e7dea">https://github.com/chromium/chromium/commit/77c8625da85bd956992b81233d451b87c39e7dea</a>

This bug is fixed in 319893@main.

* LayoutTests/svg/parser/whitespace-angle-1-expected.txt: Added.
* LayoutTests/svg/parser/whitespace-angle-1.html: Added.
* LayoutTests/svg/parser/whitespace-angle-2-expected.txt: Added.
* LayoutTests/svg/parser/whitespace-angle-2.html: Added.

Canonical link: <a href="https://commits.webkit.org/319958@main">https://commits.webkit.org/319958@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7885b44850f0729d465436df47c539c6318f7ab1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/183151 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/57074 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/50891 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/193369 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/147440 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/66728c80-af49-4f00-bb29-6566673fe4ed) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/58416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/56809 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/142953 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/99278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/831baac0-8911-4412-8014-52345701a107) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/186106 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/46681 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/163718 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/123380 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/571fc2b8-9843-4640-a4c5-35fdc31a8bb3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/45372 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/43618 "Passed tests") | | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/155770 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/43246 "Passed tests") | [✅ 🛠 gtk3-gcc](https://ews-build.webkit.org/#/builders/284/builds/4543 "Built successfully") | | 
| | | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/47881 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/195310 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/56743 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/163211 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/152434 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/40886 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/57448 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/39866 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/152129 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/46681 "Passed tests") | | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/125869 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/55956 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/18257 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/55327 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/55565 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/55394 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on LayoutTests/svg/parser/whitespace-angle-1-expected.txt:
```diff
@@ -0,0 +1,530 @@
+FAIL Test <angle> valid value:  0.35  assert_approx_equals: expected 0.35 +/- 5.960464477539063e-8 but got 0
+FAIL Test <angle> valid value:  0.35deg  assert_approx_equals: expected 0.35 +/- 5.960464477539063e-8 but got 0
+FAIL Test <angle> valid value:  0.35rad  assert_approx_equals: expected 0.35 +/- 5.960464477539063e-8 but got 0
+FAIL Test <angle> valid value:  0.35grad  assert_approx_equals: expected 0.35 +/- 5.960464477539063e-8 but got 0
```
Do you know why the code fails for these?

specifically given that similar values pass.

Ahmad-S792 replied:
Everything passes now.

karlcow commented on LayoutTests/svg/parser/whitespace-angle-1-expected.txt:
```diff
@@ -0,0 +1,530 @@
+FAIL Test <angle> valid value:  0.35\r\n\t  assert_approx_equals: expected 0.35 +/- 5.960464477539063e-8 but got 0
+FAIL Test <angle> valid value:  0.35deg\r\n\t  assert_approx_equals: expected 0.35 +/- 5.960464477539063e-8 but got 0
+FAIL Test <angle> valid value:  0.35rad\r\n\t  assert_approx_equals: expected 0.35 +/- 5.960464477539063e-8 but got 0
+FAIL Test <angle> valid value:  0.35grad\r\n\t  assert_approx_equals: expected 0.35 +/- 5.960464477539063e-8 but got 0
```
Same here. This is strange.

Ahmad-S792 replied:
Fixed - now everything passes.

karlcow commented on LayoutTests/svg/parser/whitespace-number-expected.txt:
```diff
@@ -0,0 +1,1460 @@
+PASS Test <number> trailing garbage, value: \r\n\t 1e-10)90
+FAIL Test <number> trailing garbage, value: \r\n\t +32)90 assert_approx_equals: expected 0 +/- 5.960464477539063e-8 but got 32
+FAIL Test <number> trailing garbage, value: \r\n\t +17E-1)90 assert_approx_equals: expected 0 +/- 5.960464477539063e-8 but got 1.7000000476837158
+FAIL Test <number> trailing garbage, value: \r\n\t 17e+2)90 assert_approx_equals: expected 0 +/- 5.960464477539063e-8 but got 1700
```
trailing garbage tests fail. I'm trying to understand if it's because the C++ normalize the code by dropping everything after parsing what is right. 

I also wonder what are the results in Firefox and Chrome for this test?

karlcow replied:
I wonder how these influence the values. But probably for another PR. 
https://github.com/WebKit/WebKit/blob/cfa4aa1a28ba4bf125f016ca98668f3bee093d28/Source/WebCore/svg/SVGLengthValue.cpp#L355-L357

karlcow replied:
I wonder if the tests should be separated in different categories. Just thinking.

Ahmad-S792 replied:
Tackled already.

nullhook commented on Source/WebCore/svg/SVGStopElement.cpp:
```diff
@@ -56,10 +56,36 @@ Ref<SVGStopElement> SVGStopElement::create(const QualifiedName& tagName, Documen
+            Ref { m_offset }->setBaseValInternal(*offset);
         else
-            Ref { m_offset }->setBaseValInternal(newValue.toFloat());
+            Ref { m_offset }->setBaseValInternal(0.0f); // Fallback to default value (0) for invalid input
```
is this a behavior change? the previous code used `AtomString::toFloat()`

Ahmad-S792 replied:
Done separately.

nullhook commented on Source/WebCore/svg/SVGParserUtilities.h:
```diff
@@ -32,14 +32,16 @@ namespace WebCore {
-enum class SuffixSkippingPolicy {
-    DontSkip,
-    Skip
+enum SVGWhitespaceMode {
```
why was this changed from `enum class` to `enum` ?

nullhook replied:
ideally, we can use `OptionSet<SVGWhitespaceMode>`

```c++
enum class SVGWhitespaceMode : uint8_t {
      AllowLeadingWhitespace  = 1 << 0,
      AllowTrailingWhitespace = 1 << 1,
};
```

your overloads become:

```c++
std::optional<float> parseNumber(StringParsingBuffer<Latin1Character>&, OptionSet<SVGWhitespaceMode> = { SVGWhitespaceMode::AllowLeadingWhitespace, SVGWhitespaceMode::AllowTrailingWhitespace });
```

and for checks you can use `.contains` instead of bitwise ops:

```
mode.contains(SVGWhitespaceMode::AllowLeadingWhitespace)
```

call sites can be or any of the flags

```
parseNumber(buffer, SVGWhitespaceMode::AllowLeadingWhitespace);
```

if you want `DisallowWhitespace` state then that would be an empty OptionSet:

```
parseNumber(buffer, { });
```

Ahmad-S792 replied:
Done.

Ahmad-S792 committed:
[SVG2] Allow leading and trailing whitespace in svg attributes using <angle>
https://bugs.webkit.org/show_bug.cgi?id=261656
rdar://115963075

Reviewed by NOBODY (OOPS!).

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Partial Merge: https://github.com/chromium/chromium/commit/77c8625da85bd956992b81233d451b87c39e7dea

This patch updates our implementation to match proper whitespace handling
in SVG attributes for <angle>.

Additionally, this replaces the plain enum SVGWhitespaceMode with
OptionSet<SVGWhitespaceMode> to follow WebKit conventions.

Next step would be to move `parseAngleType` to CSS Parser infrastructure
similar to `SVGLengthValue`.

* Source/WebCore/svg/SVGAngleValue.cpp:
(WebCore::parseAngleType):
(WebCore::SVGAngleValue::setValueAsString):
* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::parseKeySplines):
* Source/WebCore/svg/SVGFitToViewBox.cpp:
(WebCore::SVGFitToViewBox::parseViewBoxGeneric):
* Source/WebCore/svg/SVGParserUtilities.cpp:
(WebCore::genericParseNumber):
(WebCore::parseNumber):
(WebCore::parseNumberOptionalNumber):
(WebCore::parseRect):
* Source/WebCore/svg/SVGParserUtilities.h:
(WebCore::parseNumber):
* Source/WebCore/svg/SVGPointList.cpp:
(WebCore::SVGPointList::parse):
* Source/WebCore/svg/SVGStopElement.cpp:
(WebCore::SVGStopElement::attributeChanged):
* Source/WebCore/svg/SVGTransformListInlines.h:
(WebCore::parseTransformParamList):
* LayoutTests/svg/parser/whitespace-angle-1-expected.txt: Added.
* LayoutTests/svg/parser/whitespace-angle-1.html: Added.
* LayoutTests/svg/parser/whitespace-angle-2-expected.txt: Added.
* LayoutTests/svg/parser/whitespace-angle-2.html: Added.

nullhook commented on Source/WebCore/svg/SVGParserUtilities.h:
```diff
@@ -49,8 +50,8 @@ std::optional<bool> NODELETE parseArcFlag(StringParsingBuffer<char16_t>&);
-std::optional<FloatPoint> NODELETE parseFloatPoint(StringParsingBuffer<Latin1Character>&);
-std::optional<FloatPoint> NODELETE parseFloatPoint(StringParsingBuffer<char16_t>&);
+std::optional<FloatPoint> parseFloatPoint(StringParsingBuffer<Latin1Character>&);
+std::optional<FloatPoint> parseFloatPoint(StringParsingBuffer<char16_t>&);
```
Why was `NODELETE` removed?

Ahmad-S792 replied:
@nullhook - It was giving Safer C++ error since my function changes the use case and NODELETE annotation was no longer applicable.

nullhook replied:
The body of these functions have not changed at all. You're loosening the call-site contract by doing this.

What was the safer C++ warning about?

Ahmad-S792 replied:
We don't maintain old runs info - so I can't recall or get right now, I can add back NODELETE and see it cause Safer C++ warning.

Ahmad-S792 replied:
@nullhook - Added back - let's see.

Ahmad-S792 replied:
@nullhook - Here you go - you can see failures on safer c++ bots.

Ahmad-S792 replied:
@nullhook - with OptionSet - either I have to remove 'NODELETE' or do suppression. Or I can go back to enum class. Let me know so I can revise this and get this PR over the line.

Ahmad-S792 replied:
Based on our today's discussion, back to enum class with FIXME.

shallawa commented on Source/WebCore/svg/SVGAngleValue.cpp:
```diff
@@ -93,25 +93,28 @@ String SVGAngleValue::valueAsString() const
+        buffer += 3;
+        return SVGAngleValue::SVG_ANGLETYPE_DEG;
+    }
+    if (remaining >= 3 && compareCharacters(buffer.position(), 'r', 'a', 'd')) {
```
Why do we have to check `remaining >= 3` twice?

Ahmad-S792 replied:
Tweaked. @shallawa - if you can have a looked.

shallawa commented on Source/WebCore/svg/SVGAngleValue.cpp:
```diff
@@ -93,25 +93,28 @@ String SVGAngleValue::valueAsString() const
+        buffer += 4;
+        return SVGAngleValue::SVG_ANGLETYPE_GRAD;
+    }
+    if (remaining >= 4 && compareCharacters(buffer.position(), 't', 'u', 'r', 'n')) {
```
Can't it be just:
```
    if (remaining >= 4 && (compareCharacters(buffer.position(), 'g', 'r', 'a', 'd') || compareCharacters(buffer.position(), 't', 'u', 'r', 'n'))) {
        ...
    }
```

shallawa commented on Source/WebCore/svg/SVGParserUtilities.cpp:
```diff
@@ -43,8 +43,11 @@ template <typename FloatType> static inline bool NODELETE isValidRange(const Flo
-template <typename CharacterType, typename FloatType = float> static std::optional<FloatType> NODELETE genericParseNumber(StringParsingBuffer<CharacterType>& buffer, SuffixSkippingPolicy skip = SuffixSkippingPolicy::Skip)
+template <typename CharacterType, typename FloatType = float> static std::optional<FloatType> NODELETE genericParseNumber(StringParsingBuffer<CharacterType>& buffer, SVGWhitespaceMode mode)
 {
+    if (mode == SVGWhitespaceMode::AllowLeadingWhitespace || mode == SVGWhitespaceMode::AllowLeadingAndTrailingWhitespace)
```
```
    if (mode & SVGWhitespaceMode::AllowLeadingWhitespace)
```

Ahmad-S792 replied:
Done.

shallawa commented on Source/WebCore/svg/SVGParserUtilities.cpp:
```diff
@@ -144,26 +147,26 @@ template <typename CharacterType, typename FloatType = float> static std::option
         return std::nullopt;
 
-    if (skip == SuffixSkippingPolicy::Skip)
+    if (mode == SVGWhitespaceMode::AllowTrailingWhitespace || mode == SVGWhitespaceMode::AllowLeadingAndTrailingWhitespace)
```
Ditto.

Ahmad-S792 replied:
Done

shallawa commented on Source/WebCore/svg/SVGParserUtilities.h:
```diff
@@ -32,14 +32,17 @@ namespace WebCore {
-    DontSkip,
-    Skip
+// FIXME: Change to OptionSet in future after OptionSet and NODELETE issue is resolved.
+enum class SVGWhitespaceMode : uint8_t {
```
This name is a little bit vague. What is `SVGWhitespaceMode`? It should mention that it is about parsing white spaces rules and just white spaces.

Ahmad-S792 replied:
Done.

Ahmad-S792 replied:
I also renamed it to SVGParsingWhitespaceMode to make it clearer.

shallawa commented on Source/WebCore/svg/SVGParserUtilities.h:
```diff
@@ -32,14 +32,17 @@ namespace WebCore {
+    DisallowWhitespace                = 0,
+    AllowLeadingWhitespace            = 1 << 0,
+    AllowTrailingWhitespace           = 1 << 1,
+    AllowLeadingAndTrailingWhitespace = (1 << 0) | (1 << 1),
```
```
AllowLeadingAndTrailingWhitespace = AllowLeadingWhitespace | AllowTrailingWhitespace,
```

shallawa commented on Source/WebCore/svg/SVGPointList.cpp:
```diff
@@ -46,7 +47,7 @@ bool SVGPointList::parse(StringView value)
                 return false;
 
-            auto yPos = parseNumber(buffer, SuffixSkippingPolicy::DontSkip);
+            auto yPos = parseNumber(buffer, SVGWhitespaceMode::DisallowWhitespace);
```
The name `SuffixSkippingPolicy::DontSkip` was more obvious than `SVGWhitespaceMode::DisallowWhitespace`.  In what exactly we should `DisallowWhitespace`?

Ahmad-S792 replied:
Any suggestion for `DisallowWhitespace` rename? `NoWhitespace` or `NoSurroundingWhitespace` or `DisallowSurroundingWhitespace`?

Ahmad-S792 committed:
[SVG2] Allow leading and trailing whitespace in svg attributes using <angle>
https://bugs.webkit.org/show_bug.cgi?id=261656
rdar://115963075

Reviewed by NOBODY (OOPS!).

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Partial Merge: https://github.com/chromium/chromium/commit/77c8625da85bd956992b81233d451b87c39e7dea

This patch updates our implementation to match proper whitespace handling
in SVG attributes for <angle>.

Next step would be to move `parseAngleType` to CSS Parser infrastructure
similar to `SVGLengthValue`.

* Source/WebCore/svg/SVGAngleValue.cpp:
(WebCore::parseAngleType):
(WebCore::SVGAngleValue::setValueAsString):
* Source/WebCore/svg/SVGAnimationElement.cpp:
(WebCore::parseKeySplines):
* Source/WebCore/svg/SVGFitToViewBox.cpp:
(WebCore::SVGFitToViewBox::parseViewBoxGeneric):
* Source/WebCore/svg/SVGParserUtilities.cpp:
(WebCore::genericParseNumber):
(WebCore::parseNumber):
(WebCore::parseNumberOptionalNumber):
(WebCore::parseRect):
* Source/WebCore/svg/SVGParserUtilities.h:
(WebCore::parseNumber):
* Source/WebCore/svg/SVGPointList.cpp:
(WebCore::SVGPointList::parse):
* Source/WebCore/svg/SVGStopElement.cpp:
(WebCore::SVGStopElement::attributeChanged):
* Source/WebCore/svg/SVGTransformListInlines.h:
(WebCore::parseTransformParamList):
* LayoutTests/svg/parser/whitespace-angle-1-expected.txt: Added.
* LayoutTests/svg/parser/whitespace-angle-1.html: Added.
* LayoutTests/svg/parser/whitespace-angle-2-expected.txt: Added.
* LayoutTests/svg/parser/whitespace-angle-2.html: Added.

Ahmad-S792 committed:
[Test Coverage] [SVG2] Allow leading and trailing whitespace in svg attributes using <angle>
https://bugs.webkit.org/show_bug.cgi?id=261656
rdar://115963075

Reviewed by Taher Ali.

This patch adds test coverage for whitespace handling in SVG attributes
using <angle>, matching Gecko / Firefox and Blink / Chromium behavior.

Merge: https://github.com/chromium/chromium/commit/77c8625da85bd956992b81233d451b87c39e7dea

This bug is fixed in 319893@main.

* LayoutTests/svg/parser/whitespace-angle-1-expected.txt: Added.
* LayoutTests/svg/parser/whitespace-angle-1.html: Added.
* LayoutTests/svg/parser/whitespace-angle-2-expected.txt: Added.
* LayoutTests/svg/parser/whitespace-angle-2.html: Added.

Canonical link: https://commits.webkit.org/319958@main
